Beta CTS Docs (#12916)
hashicorp/consul: website/content/docs/release-notes/consul-terraform-sync/v0_6_0_beta.mdx

---
layout: docs
page_title: CTS v0.6.0 Beta
description: >-
  Consul-Terraform-Sync beta release notes for v0.6.0
---

-> The beta release notes for v0.6.0 focus on changes and deprecations in the upcoming Consul-Terraform-Sync (CTS) v0.6.0 release. Details of these features are subject to change as the beta period commences.

# New and Deprecated Features in v0.6.0

**Feature Summary:**

- Unify CTS UX With New CLI Start Command
- Introduction of CLI Autocompletion
- Addition of Features to Support HCP-Consul <EnterpriseAlert inline />
- Addition of Features to Support TFC Agents <EnterpriseAlert inline />

> 

Author: From @lornasong: Add a link to the changelog

Author: I've added it, thanks Lorna!

| `-config-file` &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;| Optional | string | A file to load for configuring CTS. Configuration file requires an .hcl or .json extension in order to specify their format. This option can be specified multiple times to load different configuration files. | none |
| `-inspect` | Optional | boolean | Run CTS in Inspect mode to print the proposed state changes for all tasks, and then exit. No changes are applied in this mode. | false |
| `-inspect-task` &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; | Optional | string | Run CTS in Inspect mode to print the proposed state changes for the task, and then exit. The flag can be specified multiple times to inspect multiple tasks. No changes are applied in this mode. | none |
| `-once` | Optional | boolean | Render templates and run tasks once. Does not start the process in long-running mode and disables buffer periods. | false |

## Introduction of CLI Autocompletion

> In v0.6.0 the `consul-terraform-sync` command features opt-in autocompletion for flags, subcommands, and

Reviewer: Is there a reason why we choose to do opt-in vs. enabled by default?

Author: I think it's because they have to install a separate script to handle the auto-completion. We just automate this installation. This is also the same UX as other products (Consul, Nomad, etc.)

arguments (where supported).

Enable autocompletion by running:

```shell-session
$ consul-terraform-sync -autocomplete-install
```

~> Be sure to **restart your shell** after installing autocompletion!

> When you start typing a CTS command, press the `<tab>` character to show a

Reviewer 2: ```suggestion
When you start typing a CTS command, press the `<tab>` key to show a
```

| ------------------------------------------- |-------------------------------------------- |---------------------------------- |
| 30 days before expiration                   | License retrieval attempt every 24-hours    | Warning-level log every 24-hours  |
| 7 days before expiration                    | License retrieval attempt every 1 hour      | Warning-level log every 1 hour    |
| 1 day before expiration                     | License retrieval attempt every 5 minutes   | Warning-level log every 5 minutes |
| 1 hour before expiration                    | License retrieval attempt every 1 minute    | Warning-level log every 1 minute  |
| At or after expiration (before termination) | License retrieval attempt every 1 minute    | Error-level log every 1 minute    |
| At or after termination                     | Error-level log and exit                    | Error-level log and exit          |

#### License Configuration

~> Configuring [`license_path`](/docs/nia/configuration#license_path) <EnterpriseAlert inline /> will be deprecated in v0.6.0. The described `license` block will be the preferred way to set the license configuration.

> The `license` block is used to configure how CTS attains a license. A license can be provided as a path to a license file, or CTS can be configured to automatically retrieve a license from Consul. To use the license block for CTS Enterprise, a license path must be provided or automatic license retrieval enabled. If this block is not provided, `license.auto_retrieval.enabled` will be set to true.

Author: ```suggestion
To configure how CTS attains a license, use the 'license' block. You can:
- Provide a path to a license file.
- Configure CTS to automatically retrieve a license from Consul.
If a license block is not configured, 'license.auto_retrieval.enabled' defaults to true.
```

```hcl
license {
  path = "path/to/license.lic"
  auto_retrieval {
    enabled = true
  }
}
```

##### License Config
| Parameter | Required | Type | Description | Default |
| --------- | -------- | ---- | ----------- | ------- |
| `path` |  Optional | string | Configures the path to the file containing a license. If a path to a license is configured this license will be use until, if enabled, a new license is automatically retrieved by CTS. You can also set the license by defining the `CONSUL_LICENSE` and `CONSUL_LICENSE_PATH` environment variables. See [Setting the License](/docs/nia/enterprise/license#setting-the-license) for details. | none |
| `auto_retrieval` |  Optional | object | Configures the license auto-retrieval used by CTS. See [Auto-Retrieval Config](/docs/release-notes/consul-terraform-sync/v0_6_0_beta#auto-retrieval-config) for details. | See [Auto-Retrieval Config](/docs/release-notes/consul-terraform-sync/v0_6_0_beta#auto-retrieval-config) for defaults. |

##### Auto-Retrieval Config

> The `auto_retrieval` block is used to configure the license auto-retrieval used by CTS. If enabled, CTS will attempt to retrieve a new license from its configured Consul Enterprise backend each day. In the case where a license was not able to be retrieved on the daily cadence, and the license is reaching expiry, CTS will attempt to retrieve a license at each period defined by the [License Expiry Handling](/docs/release-notes/consul-terraform-sync/v0_6_0_beta#license-expiry-handling).

Author: ```suggestion
You can use the `auto_retrieval` block to configure the the automatic license retrieval in CTS. When enabled, CTS attempts to retrieve a new license from its configured Consul Enterprise backend once a day. If CTS cannot retrieve a license and the current license is reaching its expiration date, CTS attempts to retrieve a license with increased frequency, as defined by the [License Expiration Date Handling](/docs/release-notes/consul-terraform-sync/v0_6_0_beta#license-expiry-handling).
```

Fixed attempt to attempts

> ~> `auto_retrieval` is an important feature to have enabled when connecting to HCP-Consul. HCP-Consul has a relatively short license expiry, and in order to ensure that CTS can continue to work without restart, it is important that CTS be able to dynamically attain fresh licenses from HCP-Consul.

Author: ```suggestion
~> `auto_retrieval` is an important feature to enable when you connect to HCP Consul. To ensure that CTS can work without restarting the service, it is important for CTS to have the ability to dynamically obtain fresh licenses from HCP Consul.
```

Changed attain to obtain, and fixed grammar

| Parameter | Required | Type | Description | Default |
| --------- | -------- | ---- | ----------- | ------- |
| `enabled` |  Optional | string | If set to true, enables license auto-retrieval | true |

## Addition of Features to Support TFC Agents <EnterpriseAlert inline />

In v0.6.0, CTS officially supports [TFC agents](https://www.terraform.io/cloud-docs/agents).

### Terraform Cloud Remote Workspaces

In addition to the current information on [Terraform Cloud Remote Workspaces](/docs/nia/network-drivers/terraform-cloud#remote-workspaces), to help support TFC agents, CTS introduces the [`terraform_cloud_workspace`](/docs/release-notes/consul-terraform-sync/v0_6_0_beta#terraform-cloud-workspace-configuration) configuration object to the [`Task`](/docs/nia/configuration#task) object. `terraform_cloud_workspace` is used to set options like Terraform version, execution mode, and agent pool if relevant. CTS will also use any globally set workspace configurations, specified in the driver configuration [`workspaces`](/docs/nia/configuration#workspaces).

> ~> To support TFC Agents, `terraform_cloud_workspace.execution_mode` must be set to agent execution mode.

Reviewer 3: ```suggestion
~> For a task to use Terraform Cloud agents, `terraform_cloud_workspace.execution_mode` must be set to "agent" and either `terraform_cloud_workspace.agent_pool_id` or `terraform_cloud_workspace.agent_pool_name` must be specified.
```

#### Terraform Cloud Workspace Configuration
~> Configuring [`terraform_version`](/docs/nia/configuration#terraform_version) <EnterpriseAlert inline /> will be deprecated in v0.6.0, use `terraform_cloud_workspace.terraform_version` instead.

The following configuration options are for the new `terraform_cloud_workspace` `task` configuration object which configures attributes of the Terraform Cloud workspace associated with the task. This option is only available when used with the [Terraform Cloud driver](/docs/nia/configuration#terraform-cloud-driver). For global configurations of all workspaces, see [`driver.workspaces`](/docs/nia/configuration#workspaces).

> | Name            | Required | Type    | Description                     |Default                  |

Reviewer 3: These configurations shouldn't have `-` in front of them.

- Addition of Features to Support Terraform Cloud Agents <EnterpriseAlert inline />
- Get All Tasks API Endpoint
- Health API Endpoint
- CTS Service Registration with Consul

## Introduction of CLI Start Command
~> Running CTS as a daemon without using a command is deprecated in CTS 0.6.0. The feature will be removed at a much later date in a major release. The preferred way to run CTS is to use the `start` command.

### start

In v0.6.0, the `start` command is the new and preferred way to start CTS as a daemon. This command shares the same behavior as running CTS without a command.

When CTS runs as a daemon, there is no default configuration to start CTS. You must set a configuration flag `-config-file` or `-config-dir`. For example:

```shell-session
$ consul-terraform-sync start -config-file=config.hcl
```

To review a list of available flags, use the `-help` or `-h` flag.

##### Options

The `start` command supports the following options:

| Name            | Required | Type    | Description                     |Default                  |
| --------------- | -------- | ------- | ------------------------------- | ----------------------- |

> | `-config-dir ` &nbsp; &nbsp; &nbsp;| Optional | string | A directory to load files for configuring CTS. Configuration files require an .hcl or .json file extension in order to specify their format. This option can be specified multiple times to load different directories. | none |

Reviewer 4: Suggestion to replace "Optional" to align closer with earlier text: "You must set a configuration flag `-config-file` or `-config-dir`". Feel to disregard if not helpful

You can use the `auto_retrieval` block to configure the the automatic license retrieval in CTS. When enabled, CTS attempts to retrieve a new license from its configured Consul Enterprise backend once a day. If CTS cannot retrieve a license and the current license is reaching its expiration date, CTS attempts to retrieve a license with increased frequency, as defined by the [License Expiration Date Handling](/docs/release-notes/consul-terraform-sync/v0_6_0_beta#license-expiration-handling).

~> `auto_retrieval` is an important feature to enable when you connect to HCP Consul. To ensure that CTS can work without restarting the service, it is important for CTS to have the ability to dynamically obtain fresh licenses from HCP Consul.

| Parameter | Required | Type | Description | Default |
| --------- | -------- | ---- | ----------- | ------- |
| `enabled` |  Optional | string | If set to true, enables license auto-retrieval | true |

## Additional Features to Support Terraform Cloud (TFC) Agents <EnterpriseAlert inline />

In v0.6.0, CTS officially supports [Terraform Cloud agents](https://www.terraform.io/cloud-docs/agents).

### Terraform Cloud Remote Workspaces

> To help support TFC agents, CTS introduces the [`terraform_cloud_workspace`](/docs/release-notes/consul-terraform-sync/v0_6_0_beta#terraform-cloud-workspace-configuration) configuration object to the [`Task`](/docs/nia/configuration#task) object. You can use `terraform_cloud_workspace` to set options like Terraform version, execution mode, and relevant agent pools. CTS also uses any globally set workspace configurations, specified in the driver configuration [`workspaces`](/docs/nia/configuration#workspaces).

Reviewer 4: ```suggestion
To help support TFC agents, CTS introduces the [`terraform_cloud_workspace`](/docs/release-notes/consul-terraform-sync/v0_6_0_beta#terraform-cloud-workspace-configuration) configuration to the [`task`](/docs/nia/configuration#task) configuration block. You can use `terraform_cloud_workspace` to set options like Terraform version, execution mode, and relevant agent pools. CTS also uses any globally set workspace configurations, specified in the driver configuration [`workspaces`](/docs/nia/configuration#workspaces).
```

Reviewer 4: Nit suggestion - a bit personal preference. Feel free to disregard if you disagree.

```hcl
task {
  name        = "terraform_cloud_agent_task"
  description = "execute on changes to services with names starting with web using cloud agents"
  providers   = ["my-provider"]
  module      = "path/to/services-condition-module"

  condition "services" {
    regexp     = "^web.*"
    datacenter = "dc1"
  }

  terraform_cloud_workspace {
    execution_mode = "agent"
    agent_pool_id = "apool-v9ZZW1qACRjTyzWh"
    terraform_version = "1.1.5"
  }
}
```

The configuration options are:

| Name            | Required | Type    | Description                     |Default                  |
| --------------- | -------- | ------- | ------------------------------- | ----------------------- |
| `execution_mode ` &nbsp; &nbsp; &nbsp;| Optional | string | The execution mode that determines whether to use Terraform Cloud as the Terraform execution platform. Only supports "remote" or "agent".| "remote" |

> | `agent_pool_id` &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;| Optional | string | The execution mode that determines whether to use Terraform Cloud as the Terraform execution platform. Only supports "remote" or "agent". | none |

Reviewer 4: ```suggestion
| `agent_pool_id` &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;| Optional | string | The id of the agent pool that should run the Terraform workloads. Only supported if `execution_mode` is set to "agent". Either `agent_pool_id` or `agent_pool_name` are required. `agent_pool_id` takes precedence over `agent_pool_name` if both are provided.  | none |
```

Reviewer 4: Looks like this might be a a copy & paste from `execution_mode `

| `request_id` | string        | The ID of the request. Used for auditing and debugging purposes.                    |
| `tasks`      | list[[Task](/docs/nia/api/tasks#task-object)]  | A list of Tasks  |

### Example: Retrieve all tasks

In this example, CTS contains a single task

Request:

```shell-session
$ curl --request GET \
  localhost:8558/v1/tasks
```

Response:

```json
{
  "request_id": "0e0290f9-94df-3a4a-fd17-72467a16083c",
  "tasks": [
    {
      "buffer_period": {
        "enabled": true,
        "max": "25s",
        "min": "5s"
      },
      "condition": {
        "services": {
          "cts_user_defined_meta": {},
          "datacenter": "",
          "filter": "",
          "names": ["api", "web"],
          "namespace": "",
          "use_as_module_input": true
        }
      },
      "description": "Writes the service name, id, and IP address to a file",
      "enabled": true,
      "module": "path/to/module",
      "module_input": {},
      "name": "task_a",
      "providers": ["my-provider"],
      "variables": {},
      "version": ""
    }
  ]
}
```

## Health Endpoint

> The `/health` endpoint returns a successful response if Consul-Terraform-Sync (CTS) is available and running. Requests to this endpoint are not logged, which makes it suitable for health checks that constantly poll CTS.

Reviewer 4: ```suggestion
The `/health` endpoint returns a successful response if CTS is available and running. Requests to this endpoint are not logged, which makes it suitable for health checks that constantly poll CTS.
```

Reviewer 4: Suggestion to remove "Consul-Terraform-Sync" since the abbreviation is already established earlier on this page

| Method | Path                | Produces           |
| ------ | ------------------- | ------------------ |
| `GET`  | `/health` | `application/json` |

### Response Statuses

| Status | Reason                                               |
| ------ | ---------------------------------------------------- |
| 200    | CTS is healthy                                       |

### Response Fields

| Name         | Type   | Description                                                                        |
| ------------ | ------ | ---------------------------------------------------------------------------------- |
| `request_id` | string | The ID of the request. Used for auditing and debugging purposes.                   |

### Example

The following request makes a `GET` call to the `health` endpoint:

```shell-session
$ curl --request GET \
  localhost:8558/v1/health
```

Response:

```json
{
  "request_id": "b7559ab0-5111-381b-367a-0dfb7e216d41"
}
```

## CTS Service Registration With Consul

> By default, CTS automatically registers itself as a service in Consul with a health check. The default health check is an HTTP check to the [health API endpoint](#health-endpoint). CTS deregisters itself with Consul when it is stopped.

Reviewer 4: ```suggestion
By default, CTS automatically registers itself as a service in Consul with a health check. The default health check is an HTTP check to the [health API endpoint](#health-endpoint). CTS deregisters itself with Consul when CTS stops gracefully.
```

Reviewer 4: Suggestion to clarify if "it" was referring to Consul vs. CTS. Feel free to disregard if this isn't true
